### PR TITLE
Add the random fixed size exemplar reservoir

### PR DESCRIPTION
**Pull request by MrAlias**
Part of #559

This PR is split from #4455

This adds a `Reservoir` implementation that will randomly sample a specified number of measurements as exemplars. An explanation of the algorithm is copied from the included comment within the PR:

> The following algorithm is "Algorithm L" from Li, Kim-Hung (4 December 1994). "Reservoir-Sampling Algorithms of Time Complexity O(n(1+log(N/n)))". ACM Transactions on Mathematical Software. 20 (4): 481–493 (https://dl.acm.org/doi/10.1145/198429.198435).
> 
> A high-level overview of "Algorithm L":
>
>   0) Pre-calculate the random count greater than the storage size when an exemplar will be replaced.
>   1) Accept all measurements offered until the configured storage size is reached.
>   2) Loop:
>      a) When the pre-calculate count is reached, replace a random existing exemplar with the offered measurement.
>      b) Calculate the next random count greater than the existing one which will replace another exemplars
> 
> The way a "replacement" count is computed is by looking at `n` number of independent random numbers each corresponding to an offered measurement. Of these numbers the smallest `k` (the same size as the storage capacity) of them are kept as a subset. The maximum value in this subset, called `w` is used to weight another random number generation for the next count that will be considered.
> 
> By weighting the next count computation like described, it is able to perform a uniformly-weighted sampling algorithm based on the number of samples the reservoir has seen so far. The sampling will "slow down" as more and more samples are offered so as to reduce a bias towards those offered just prior to the end of the collection.
> 
> This algorithm is preferred because of its balance of simplicity and performance. It will compute three random numbers (the bulk of computation time) for each item that becomes part of the reservoir, but it does not spend any time on items that do not. In particular it has an asymptotic runtime of O(k(1 + log(n/k)) where n is the number of measurements offered and k is the reservoir size.
> 
> See https://en.wikipedia.org/wiki/Reservoir_sampling for an overview of this and other reservoir sampling algorithms. See https://github.com/MrAlias/reservoir-sampling for a performance comparison of reservoir sampling algorithms.

### CODE REVIEW

File: sdk/metric/internal/exemplar/rand.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+	// The following algorithm is "Algorithm L" from Li, Kim-Hung (4 December
+	// 1994). "Reservoir-Sampling Algorithms of Time Complexity
+	// O(n(1+log(N/n)))". ACM Transactions on Mathematical Software. 20 (4):
+	// 481–493 (https://dl.acm.org/doi/10.1145/198429.198435).
+	//
+	// It is used because of its balance of simplicity and performance. In
+	// particular it has an asymptotic runtime of O(k(1 + log(n/k)) where n is
+	// the number of measurements offered and k is the reservoir size. This is
+	// much more optimal for large measurement sets than the algorithm
+	// recommended by the OTel spcification ("Algorithm R" as described in
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/rand.go:66]: Should we update the spec to recommend this algorithm instead?  Since it is just as an example algorithm, i'm OK to use this one instead.

File: sdk/metric/internal/exemplar/rand.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	r.reset()
+	return r
+}
+
+type randRes[N int64 | float64] struct {
+	*storage[N]
+
+	// count is the number of measurement seen.
+	count int64
+	// next is the next count that will store a measurement at a randon index
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/rand.go:48]:
```suggestion
	// next is the next count that will store a measurement at a random index
```

File: sdk/metric/internal/exemplar/storage.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		n++
+
+		// Reset.
+		r.store[i] = measurement[N]{}
+	}
+	*dest = (*dest)[:n]
+}
+
+// measurement is a measurement made by a telemetry system.
+type measurement[N int64 | float64] struct {
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/storage.go:79]: A comment explaining why we don't re-use metricdata.Exemplar would be helpful.  I assume it is to move the conversion of SpanContext to TraceID and SpanID to the Collect, and keep it out of the hot path, but it would be helpful to document that.

File: sdk/metric/internal/exemplar/rand.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+			r.advance()
+		}
+	}
+	r.count++
+}
+
+func (r *randRes[N]) reset() {
+	r.count = 0
+	r.next = int64(cap(r.store))
+	r.w = math.Exp(math.Log(rng.Float64()) / float64(cap(r.store)))
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/rand.go:89]: Looks like this is already part of advance().  Is this line still needed?

File: sdk/metric/internal/exemplar/rand.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	// runtime of O(n).
+	//
+	// See https://github.com/MrAlias/reservoir-sampling for a comparison of
+	// reservoir sampling algorithms (including performance benchmarks).
+
+	if int(r.count) < cap(r.store) {
+		r.store[r.count] = newMeasurement(ctx, t, n, a)
+	} else {
+		if r.count == r.next {
+			idx := int(rng.Int63n(int64(cap(r.store))))
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/rand.go:78]:
```suggestion
                         // Overwrite a random existing measurement with the new one.
			idx := int(rng.Int63n(int64(cap(r.store))))
```

File: sdk/metric/internal/exemplar/rand.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+func (r *randRes[N]) reset() {
+	r.count = 0
+	r.next = int64(cap(r.store))
+	r.w = math.Exp(math.Log(rng.Float64()) / float64(cap(r.store)))
+	r.advance()
+}
+
+func (r *randRes[N]) advance() {
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/rand.go:93]: I know you link to the original source of the algorithm above, but can you summarize or explain in a sentence or two how the algorithm works? I assume it "slows down" sampling over time to reduce bias towards measurements offered just before Collect.

File: sdk/metric/internal/exemplar/rand.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	r.w *= math.Exp(math.Log(rng.Float64()) / float64(cap(r.store)))
+	r.next += int64(math.Log(rng.Float64())/math.Log(1-r.w)) + 1
+}
+
+func (r *randRes[N]) Collect(dest *[]metricdata.Exemplar[N]) {
+	r.storage.Collect(dest)
+	r.reset()
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/rand.go:100]: If collect isn't supposed to clear data, I wouldn't expect it to reset here.  Won't it just immediately overwrite all of the existing exemplars then, making it roughly equivalent to Flush?

File: sdk/metric/internal/exemplar/rand.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:66]: I have already updated the specification to allow this algorithm (https://github.com/open-telemetry/opentelemetry-specification/pull/3627) and remove the recommendation on the more naive implementation. I'll update the comment to match.

I left the specification example in place though. Given that is the easiest example to show in the spec I thought it fine given it was no longer a recommendation to use it. Do you think the spec still needs to be updated?

File: sdk/metric/internal/exemplar/storage.go (hunk 3 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/storage.go:79]: This is documented here: https://github.com/open-telemetry/opentelemetry-go/pull/4852/files/51da2b9d75875c62e198786f601542f080a9063a#diff-51b6936ff08d90e60dec222a0728ea140a31fd4753a0c995046b709bf94cd170R30-R31

Are you saying we should document this in both places?

File: sdk/metric/internal/exemplar/rand.go (hunk 6 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:93]: Sure. I can give it a try.

File: sdk/metric/internal/exemplar/rand.go (hunk 7 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:100]: This comes from the [specification](https://github.com/MrAlias/opentelemetry-specification/blob/main/specification/metrics/sdk.md#simplefixedsizeexemplarreservoir):

> Any stateful portion of sampling computation SHOULD be reset every collection
> cycle. For the above example, that would mean that the `num_measurements_seen`
> count is reset every time the reservoir is collected.

Your are correct in that if more measurements are made in the next collection cycle this will be equivalent to a flush. But I think the specification was trying to handle two things:

- If no measurements are made in the next collection cycle persist old exemplars
- If there are measurements made in the next collection cycle, prioritize those as exemplars.

In the end this method will be different from `Flush` in that it will always return `n` exemplars once `n` exemplars have been recorded.

File: sdk/metric/internal/exemplar/rand.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go]: Recommendation comment removed: https://github.com/open-telemetry/opentelemetry-go/pull/4852/commits/498f3d721d10d4d2ce856484963ad597e47eb9c7

File: sdk/metric/internal/exemplar/rand.go (hunk 4 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:84]: There's a slight difference here. This is initializing `r.w` (i.e. the needed starting point of the random number series https://dl.acm.org/doi/pdf/10.1145/198429.198435), but in `advance` it is determining the next random number in index to make the insert. Specifically, it is doing `r.w = r.w * math.Exp(...)`.

If this isn't reset here:

- Initial values of `r.w` will be `0`, so all subsequent values of `r.w` will also be `0`
- When reset is called during a collection `r.w` will continue to grow in the last random number set instead of restarting another

Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:84]: I'll add a comment to clarify

Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:93]: commented: https://github.com/open-telemetry/opentelemetry-go/pull/4852/commits/b8af1e7dccea17c8af203133f0abe85ecb8b4194

File: sdk/metric/internal/exemplar/rand.go (hunk 7 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:114]: commented: https://github.com/open-telemetry/opentelemetry-go/pull/4852/commits/b8af1e7dccea17c8af203133f0abe85ecb8b4194

File: sdk/metric/internal/exemplar/rand.go (hunk 5 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:74]: https://github.com/open-telemetry/opentelemetry-go/pull/4852/commits/1150d3d64f23b1b4b583b2cc36e9260e79928918

File: sdk/metric/internal/exemplar/rand.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	r.w = math.Exp(math.Log(rng.Float64()) / float64(cap(r.store)))
+
+	// Calculate the next random index a measurement will become an exemplar.
+	r.advance()
+}
+
+// advance updates the random number series of r and the next insert index.
+func (r *randRes[N]) advance() {
+	// Next value in the random number series based on the existing r.w.
+	r.w *= math.Exp(math.Log(rng.Float64()) / float64(cap(r.store)))
```
Comment (MrAlias) [inline sdk/metric/internal/exemplar/rand.go:103]: `rng.Float64` is defined on the half-open interval `[0, 1]`. This is not valid for `0` and it needs to be refactored to avoid this error.

File: sdk/metric/internal/exemplar/storage.go (hunk 3 above)
Comment (dashpole) [reply sdk/metric/internal/exemplar/storage.go:79]: Ah, I missed that.  I'm OK with not duplicating it here.

File: sdk/metric/internal/exemplar/rand.go (hunk 4 above)
Comment (dashpole) [reply sdk/metric/internal/exemplar/rand.go:94]: LGTM, thanks

File: sdk/metric/internal/exemplar/rand.go (hunk 7 above)
Comment (dashpole) [reply sdk/metric/internal/exemplar/rand.go:115]: thanks.  Makes sense.

File: sdk/metric/internal/exemplar/rand.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	// Nishimura and Makoto Matsumoto:
+	// http://www.math.sci.hiroshima-u.ac.jp/m-mat/MT/VERSIONS/C-LANG/mt19937-64.c)
+	//
+	//   (float64(rng.Int63()>>11) + 0.5) * (1.0 / 4503599627370496.0)
+	//
+	// There are likely many other methods to explore here as well.
+
+	f := rng.Float64()
+	for f == 0 {
+		f = rng.Float64()
```
Comment (dashpole) [inline sdk/metric/internal/exemplar/rand.go:55]: can't this still be zero?

File: sdk/metric/internal/exemplar/rand.go (hunk 6 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:161]: added in https://github.com/open-telemetry/opentelemetry-go/pull/4852/commits/fef8df6cf8f6fa77c84f03c3e9443e3f40f4578a. PTAL

File: sdk/metric/internal/exemplar/rand.go (hunk 9 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:55]: It is looped until `f` is not zero.

Talked about in SIG meeting. Therefore, going to resolve. Please re-open if done in error.

File: sdk/metric/internal/exemplar/rand.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+// rng is used to make sampling decisions.
+//
+// Do not use crypto/rand. There is no reason for the decrease in performance
+// given this is not a security sensitive decision.
+var rng = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+// random returns, as a float64, a uniform pseudo-random number in the open
+// interval (0.0,1.0).
+func random() float64 {
+	// TODO: This does not return a uniform number. rng.Float64 returns
```
Comment (MrAlias) [inline sdk/metric/internal/exemplar/rand.go:36]:
```suggestion
	// TODO: This does not return a uniform number. rng.Float64 returns a
```

File: sdk/metric/internal/exemplar/rand_test.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+		r.Offer(context.Background(), staticTime, value, nil)
+	}
+
+	var sum float64
+	for _, m := range r.(*randRes[float64]).store {
+		sum += m.Value
+	}
+	mean := sum / float64(sampleSize)
+
+	assert.InDelta(t, 1/mean, intensity, 0.01)
```
Comment (pellared) [inline sdk/metric/internal/exemplar/rand_test.go:58]: This test is flaky

```
$ go test -run=TestFixedSizeSamplingCorrectness -count=1000
--- FAIL: TestFixedSizeSamplingCorrectness (0.18s)
    rand_test.go:58: 
                Error Trace:    /home/rpajak/repos/opentelemetry-go/sdk/metric/internal/exemplar/rand_test.go:58
                Error:          Max difference between 0.11328473723564143 and 0.1 allowed is 0.01, but difference was 0.013284737235641428
                Test:           TestFixedSizeSamplingCorrectness
```

I hope that changing `sampleSize` to `10000` will be enough to make it stable. We can also consider making the delta bigger.

File: sdk/metric/internal/exemplar/rand_test.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+	t.Run("Int64", ReservoirTest[int64](func(n int) (Reservoir[int64], int) {
+		return FixedSize[int64](n), n
+	}))
+
+	t.Run("Float64", ReservoirTest[float64](func(n int) (Reservoir[float64], int) {
+		return FixedSize[float64](n), n
+	}))
+}
+
+func TestFixedSizeSamplingCorrectness(t *testing.T) {
```
Comment (pellared) [inline sdk/metric/internal/exemplar/rand_test.go:36]: First of all, I want to share that I am impressed with this unit test 👍 

This test needs description (probability theory / statistics calculations?) on how does this tests checks the correctness.

File: sdk/metric/internal/exemplar/rand.go (hunk 9 above)
Comment (pellared) [reply sdk/metric/internal/exemplar/rand.go:55]: My guess is this is to avoid getting `math.Log(0.0)`.

> Talked about in SIG meeting.

(I was not following as I was not familiar with the PR) 🙈 

Still I think that the reason should be commented so that someone would not remove this condition as it is not obvious.

However, shouldn't' we do this instead:

```go
if f == 0 {
  return 1.0
}
```

so that the function returns a pseudo-random number in the half-open interval `(0.0,1.0]`?

Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand.go:55]: No. The implemented algorithm requires the random number to be generate over the open interval `(0,1)`, as this function is documented to produce.

File: sdk/metric/internal/exemplar/rand_test.go (hunk 12 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand_test.go:36]: This is creating an exponential distribution of data which is sampled by the reservoir. The mean of the sampled data is evaluated to match the original distribution to show the sampling does not alter the distribution (i.e. no bias towards the start or end).

I can add this description as a comment, what do you think?

I'm not sure adding much more would be helpful. Testing statistical properties is a complex topic and I'm not sure providing a lecture on the subject in the comments is going to help.

File: sdk/metric/internal/exemplar/rand_test.go (hunk 11 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/rand_test.go:61]: This is looking for a delta of `0.01` which is 10*sigma. I'll relax the delta to be `0.02` (i.e. 5*sigma) which should still be a high enough confidence interval without the noise.

As for scaling the sample size, given it scales the standard deviation of the mean inverse to its square root I think we will be better served lowering the confidence interval instead (i.e. its already ~0.03*sigma).

File: sdk/metric/internal/exemplar/rand_test.go (hunk 12 above)
Comment (pellared) [reply sdk/metric/internal/exemplar/rand_test.go:36]: I find the description very helpful and good enough 👍

> providing a lecture on the subject in the comments is going to help.

This is not needed. Describing what statistical properties are used for testing is good enough. Someone who does not understand the statistics can learn further on his own.

File: sdk/metric/internal/exemplar/rand_test.go (hunk 11 above)
Comment (pellared) [reply sdk/metric/internal/exemplar/rand_test.go]: And it will also make the test execution faster 😉

File: sdk/metric/internal/exemplar/storage.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+// Collect returns all the held exemplars.
+//
+// The Reservoir state is preserved after this call. See Flush to
+// copy-and-clear instead.
+func (r *storage[N]) Collect(dest *[]metricdata.Exemplar[N]) {
+	*dest = reset(*dest, len(r.store), len(r.store))
+	var n int
+	for _, m := range r.store {
+		if !m.Valid() {
```
Comment (pellared) [inline sdk/metric/internal/exemplar/storage.go:47]: nit: I propose to change `Valid` to `Invalid` as all calls are `if !m.Valid()`. I find `if m.Invalid()` slightly more readable.

File: sdk/metric/internal/exemplar/storage.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+			continue
```
Comment (pellared) [inline sdk/metric/internal/exemplar/storage.go:48]: I think it would be nice to have this line covered. Maybe by enhancing the existing `ReservoirTest`?

**Review by pellared — APPROVED** [review]

File: sdk/metric/internal/exemplar/storage.go (hunk 13 above)
Comment (pellared) [reply sdk/metric/internal/exemplar/storage.go:47]: When I look at it again I am not sure if we even need a method.

Maybe we could simply have `if !m.valid`? 😉

File: sdk/metric/internal/exemplar/storage.go (hunk 14 above)
Comment (MrAlias) [reply sdk/metric/internal/exemplar/storage.go:48]: 001bdc81

**Review by pellared — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

**Review by hanyuancheung — APPROVED** [review]

**Review by dashpole — APPROVED** [review]
